File: peg-game-solver.c
/******************************************************************************
* peg-game-solver
* Solves the peg game described in https://github.com/chrisdana/watch-peg-game
* with peg positions ordered from top to bottom, left to right (shown below).
* Game boards with 4, 5, or 6 rows are supported.
* 
*                 0
*              1     2
*           3     4      5
*        6     7      8     9
*     10    11    12    13    14
*  15    16    17    18    19    20
*/

#include <stdio.h>
#include <stdlib.h>
# include <math.h>

#define MAX_NEIGHBORS 6
#define EMPTY -1


static int triangular_number(int n) {
    return (n * (n + 1) / 2);
}

static void add_directional_edge(int **graph, int n1, int n2) {
    int i;
    for (i = 0; i < MAX_NEIGHBORS; i++) {
        if (graph[n1][i] == EMPTY) {
            graph[n1][i] = n2;
            break;
        }
    }    
}

static void add_edge(int **graph, int n1, int n2) {
    add_directional_edge(graph, n1, n2);
    add_directional_edge(graph, n2, n1);
}

static int** gen_triangle_graph(int n_rows) {
    int n = triangular_number(n_rows);
    int i, j;
    int current_node;

    /* Allocate memory and initialize */
    int **g = malloc(n * sizeof(int*));
    for (i = 0; i < n; i++) {
        g[i] = malloc(MAX_NEIGHBORS * sizeof(int));
        for (j = 0; j < MAX_NEIGHBORS; j++) {
            g[i][j] = EMPTY;
        }
    }
    
    /* Fill in neighbors */
    for (i = 0; i < n_rows; i++) {
        for (j = 0; j < (i + 1); j++) {
            /* The first node in each row is the triangular number for that row */
            current_node = triangular_number(i) + j;
            if (i < (n_rows - 1)) {
                add_edge(g, current_node, current_node + i + 1); /* Edge ot lower-left */
                add_edge(g, current_node, current_node + i + 2); /* Edge to lower-right */
            }
            if (j < i) {
                add_edge(g, current_node, current_node + 1); /* Edge to right */
            }
        }
    }

    return g;
}

static int row_from_node(int n) {
    /* The ith row is the positive, floored solution to the equation i^2 + i - 2n = 0 */
    return (int)((-1 + sqrtf(1 + (4 * (2 * n)))) / 2);
}

static int has_peg(int n, uint32_t state) {
    return ((state >> n) & 0x01);
}

static void set_peg(int n, uint32_t *state) {
    *state = *state | (0x01 << n);
}

static void rem_peg(int n, uint32_t *state) {
    *state = *state & ~(0x01 << n);
}

static int enc_move(int src, int mid, int dest) {
    return (10000 * src) + (100 * mid) + dest;
}

static void dec_move(int move, int *src, int *mid, int *dest) {
    *src = move / 10000;
    *mid = (move / 100) % 100;
    *dest = move % 100;
}

static int count_pegs(uint32_t i) {
    int count = 0;

    while (i) {
        count += i & 1;
        i >>= 1;
    }
    
    return count;
}

static void print_bs(uint32_t bs, int n_nodes, int n_rows) {
    int next_row_idx = 0;
    int next_row_len = 1;
    printf("Board state (0 - Hole, 1 - Peg):");
    for (int i = 0; i < n_nodes; i++) {
        if (i == next_row_idx) {
            printf("\n");
            printf("%.*s", (n_rows - next_row_len), "          ");
            next_row_idx += next_row_len;
            next_row_len += 1;
        }
        printf("%d ", has_peg(i, bs));
    }
    printf("\n");
}

static int is_neighbor(int n, int k, int **graph) {
    for (int i = 0; i < MAX_NEIGHBORS; i++) {
        if (graph[n][i] == k) {
            return 1;
        }
        else if (graph[n][i] == EMPTY) {
            break;
        }
    }
    return 0;
}

/* Returns all valid moves of a given board state in *moves (if not NULL) */
static int get_valid_moves(int **graph, uint32_t state, int n_nodes, int moves[]) {
    int src, mid, dest;
    int i;
    int src_row, mid_row, dest_row;
    int n_moves = 0;
    int i_valid_moves = 0;

    for (src = 0; src < n_nodes; src++) {
        if (!(has_peg(src, state))) {
            continue;
        }

        for (i = 0; i < MAX_NEIGHBORS; i++) {
            if (graph[src][i] == EMPTY) {
                break;
            }

            mid = graph[src][i];

            if (mid == src) {
                continue;
            }

            if (!(has_peg(mid, state))) {
                continue;
            }

            src_row = row_from_node(src);
            mid_row = row_from_node(mid);

            // There are two cases: Jumping between levels and jumping
            // horizontally in the same level
            if (src_row != mid_row) {
                dest = (2 * mid) - src + 1;
            }
            else {
                dest = (2 * mid) - src;
            }

            // Dest must be positive
            if (dest < 0) {
                continue;
            }

            // Dest must be a neighbor of mid
            if (!(is_neighbor(mid, dest, graph))) {
                continue;
            }

            // Dest cannot be a neighbor of src (avoids invalid 4, 2, 1 move)
            if (is_neighbor(src, dest, graph)) {
                continue;
            }

            // Ensure horizontal jumps are on the same level
            dest_row = row_from_node(dest);
            if ((src_row == mid_row) && (mid_row != dest_row)) {
                continue;
            }

            // Dest must not have a peg
            if (has_peg(dest, state)) {
                continue;
            }

            n_moves++;
            if (moves != NULL) {
                int encoded = enc_move(src, mid, dest);
                //printf("Adding move: %d %d %d  (%d)\n", src, mid, dest, encoded);
                moves[i_valid_moves] = encoded;
                i_valid_moves++;
            }
        }
    }
    return n_moves;
}

static int n_valid_moves(int **graph, uint32_t state, int n_nodes) {
    return get_valid_moves(graph, state, n_nodes, NULL);
}

static int solve(int **graph, uint32_t curr_bs, int n_nodes, int *final_moves, int *idx_final_moves) {
    int n_moves = 0;
    int move;
    uint32_t bs = curr_bs;
    int src, mid, dest;

    /* If we have one peg left, we are done */
    if (count_pegs(bs) == 1) {
        return 1;
    }

    /* If there are valid moves, return */
    n_moves = n_valid_moves(graph, bs, n_nodes);
    if (n_moves == 0) {
        return 0;
    }

    /* Get all valid moves */
    int *moves = malloc(n_moves * sizeof(int));
    get_valid_moves(graph, bs, n_nodes, moves);

    /* Depth-first search */
    for (int i = 0; i < n_moves; i++) {
        // Reset the board state to initial function call
        bs = curr_bs;

        // Make the move
        move = moves[i];
        dec_move(move, &src, &mid, &dest);
        rem_peg(src, &bs);
        rem_peg(mid, &bs);
        set_peg(dest, &bs);

        // Add the move to the final move list
        final_moves[*idx_final_moves] = move;
        (*idx_final_moves)++;

        if (solve(graph, bs, n_nodes, final_moves, idx_final_moves) == 1) {
            return 1;
        } else {
            // Remove the last move
            (*idx_final_moves)--;
            final_moves[*idx_final_moves] = 0;
        }
    }

    // No solutions down this branch
    free(moves);
    return 0;
}

int main(int argc, char **argv) {
    int n_rows = 0;
    if (argc > 1) {
        n_rows = atoi(argv[1]);
    }

    if (n_rows < 4 || n_rows > 6) {
        fprintf(stderr, "Error: Argument invalid.\n");
        fprintf(stderr, "First argument must be number of rows in the triangle\n");
        fprintf(stderr, "The number of rows must be in the range 4-6\n");
        fprintf(stderr, "Example: ./peg-game-solver 6\n");
        exit(1);
    }

    int n_nodes = triangular_number(n_rows);
    int **graph = gen_triangle_graph(n_rows);
    int *final_moves = malloc(n_nodes * sizeof(int));
    int idx_final_moves = 0;
    int src, mid, dest;
    int curr_node;
    uint32_t init_bs;
    int ret = 0;

    // Set the initial board state
    for (int i = 0; i < (n_rows / 2) + 1; i++) {
        for (int j = 0; j < (((i + 1) / 2) + 1); j++) {
            curr_node = triangular_number(i) + j;

            printf("Trying initial state with peg %d removed\n", curr_node);
            init_bs = 0;
            for (int k = 0; k < n_nodes; k++) {
                set_peg(k, &init_bs);
            }
            rem_peg(curr_node, &init_bs);
            print_bs(init_bs, n_nodes, n_rows);

            ret = solve(graph, init_bs, n_nodes, final_moves, &idx_final_moves);
            if (ret == 1) {
                printf("Solution:\n");
                for (int i = 0; i < (n_nodes - 2); i++) {
                    dec_move(final_moves[i], &src, &mid, &dest);
                    printf("Move %d:  %d --> %d\n", (i + 1), src, dest);
                }
                free(final_moves); 
                exit(0);
            }
            printf("No solution found from this starting position.\n\n");
        }
    }

    printf("Unable to solve puzzle.\n");
    free(final_moves);
    exit(0);
}

File: peg-game-solver.py
################################################################################
# peg-game-solver.py
# Solves the peg game described in https://github.com/chrisdana/watch-peg-game
# with peg positions ordered from top to bottom, left to right (shown below).
# Game boards with 4, 5, or 6 rows are supported.
# 
#                 0
#              1     2
#           3     4      5
#        6     7      8     9
#     10    11    12    13    14
#  15    16    17    18    19    20
#
################################################################################

import networkx as nx
import sys

def triangular_number(n: int) -> int:
    """
    Computes the triangular number T(n).
    
    Parameters:
        n (int): The integer for which triangular number is to be calculated.
        
    Returns:
        int: Triangular number T(n).
    """
    return n * (n + 1) // 2

def generate_triangle_graph(n: int) -> nx.Graph:
    """
    Generates a graph representing T(n) objects arranged in an equilateral triangle.
    
    Parameters:
        n (int): The number of rows in the equilateral triangle.
        
    Returns:
        nx.Graph: Graph representing the equilateral triangle.
    """
    graph = nx.Graph()
    vertices = triangular_number(n)
    levels = []
    
    # Add vertices
    for i in range(vertices):
        graph.add_node(i)
    
    # Add edges
    for i in range(n):
        for j in range(i + 1):
            current_node = i * (i + 1) // 2 + j
            # Save the level this node is at
            graph.nodes[current_node]["level"] = i
            if i < n - 1:
                graph.add_edge(current_node, current_node + i + 1)  # Edge to bottom-left node
                graph.add_edge(current_node, current_node + i + 2)  # Edge to bottom-right node
            if j < i:
                graph.add_edge(current_node, current_node + 1)      # Edge to right node
    
    return graph


def get_valid_moves(graph: nx.Graph, bs: [bool]) -> [(int, int, int)]:
    """
    Returns an array of all valid moves of the current board state
    
    Parameters:
        graph (nx.Graph): Input graph.
        bs (list): The current board state as an array of True/False
        
    Returns:
        list: List containing all valid moves (src, peg_to_remove, dest)
    """
    moves = []
    for src in range(len(list(graph.nodes))):
        has_peg = bs[src]
        if has_peg:
            # A valid move is if a neighbor has a peg, but the next hole 
            # in the same direction is empty.
            for mid in list(graph.neighbors(src)):
                # Avoid cycles
                if mid == src:
                    continue
                
                # Neighbor must have a peg
                if not bs[mid]:
                    continue
                
                src_lvl = graph.nodes[src]["level"]
                mid_lvl = graph.nodes[mid]["level"]
                
                # There are two cases: Jumping between levels and jumping
                # horizontally in the same level
                if src_lvl != mid_lvl:
                    dest = (2 * mid) - src + 1
                else:
                    dest = (2 * mid) - src
                
                # Dest must be valid
                if dest not in graph.neighbors(mid):
                    continue
                
                # Dest cannot be a neighbor of src (avoids invalid (4, 2, 1) move)
                if dest in graph.neighbors(src):
                    continue
                
                # Ensure horizontal jumps are actually on the same level
                if src_lvl == mid_lvl and mid_lvl != graph.nodes[dest]["level"]:
                    continue
                
                # Dest must not have a peg
                if bs[dest]:
                    continue
                
                moves.append((src, mid, dest))
    return moves
            
def solve(graph: nx.Graph, bs: [bool]) -> bool:
    """
    Tries to solve a given triangle puzzle by depth-first search
    
    Parameters:
        graph (nx.Graph): Input graph.
        bs (list): The current board state (peg locations) as an array of True/False
        
    Returns:
        True if successful, False if not
        If successful the FINAL_MOVE_LIST will contain the solution
    """
    global FINAL_MOVE_LIST
    global N_MOVES
    global SEEN
    
    # If we have one peg left, we are done
    if bs.count(True) == 1:
        return True
    
    moves = get_valid_moves(graph, bs)
    
    # If there are no valid moves, return
    if len(moves) == 0:
        return False
    
    # Depth-first search 
    for m in moves:
        N_MOVES += 1
        new_bs = bs.copy()
        
        # Make the move indicated 
        new_bs[m[0]] = False
        new_bs[m[1]] = False
        new_bs[m[2]] = True
        
        # Encode board state in hashable form 
        bs_hash = ''.join(['1' if x else '0' for x in new_bs])
        
        # Skip if we have seen this board before
        if bs_hash in SEEN:
            continue
        
        SEEN.add(bs_hash)
        FINAL_MOVE_LIST.append(m)
        
        if solve(graph, new_bs) == True:
            return True
        else:
            FINAL_MOVE_LIST.pop()
                
    # Valid moves led to board states we have already seen
    return False

def print_bs(bs: [bool], n_rows: int) -> None:
    """
    Prints the board state

    Parameters:
        bs ([Bool]): List of boolean values representing peg locations
        n_rows (Int): Number of rows in the triangle
        
    Returns:
        None
    """
    next_row_idx = 0
    next_row_len = 1
    print("Board state (X - Hole, O - Peg):")
    for i in range(len(bs)):
        if i == next_row_idx:
            print("")
            print(" " * (n_rows - next_row_len), end="")
            next_row_idx += next_row_len
            next_row_len += 1
        print("%s " %("O" if bs[i] else "X"), end="")
    print("\n")
        
def usage() -> None:
    print("Error: Argument invalid.", file=sys.stderr)
    print("First argument must be number of rows in the triangle", file=sys.stderr)
    print("The number of rows must be in the range 4-6", file=sys.stderr)
    print("Example: python3 peg-game-solver.py 6", file=sys.stderr)

def main() -> int:
    global FINAL_MOVE_LIST
    global N_MOVES
    global SEEN

    FINAL_MOVE_LIST = []
    N_MOVES = 0
    SEEN = set()

    # Number of levels in the triangle (only works for N = [4,5,6])
    try:
        n_rows = int(sys.argv[1])
        if n_rows < 4 or n_rows > 6:
            raise
    except Exception:
        usage()
        sys.exit(1)
    
    triangle_graph = generate_triangle_graph(n_rows)
      
    # Initial board states can be reduced due to symmetry
    # We only need check (n_rows // 2) + 1 levels 
    # In each level we only need check (Xi // 2) + 1 nodes if Xi is the number of nodes in the ith level
    for i in range((n_rows // 2) + 1):
        for j in range(((i + 1) // 2) + 1):
            current_node = i * (i + 1) // 2 + j
            
            print("Trying inital board state with peg %d removed" % current_node)
            # Remove the initial peg and run the solver
            board_state = [True] * len(triangle_graph.nodes)
            board_state[current_node] = False
            print_bs(board_state, n_rows)
            
            if solve(triangle_graph, board_state):
                print("Solution:")
                for i, move in enumerate(FINAL_MOVE_LIST):
                    print("Move %d: %d --> %d" %((i + 1), move[0], move[2]))
                sys.exit(0)
            
            print("No solution found from this starting position.\n")

    print("Unable to solve puzzle. ")
    sys.exit(0)

if __name__ == '__main__':
    main()
